core/src/main/scala/org/apache/spark/ui/HttpSecurityFilter.scala

```diff
@@ -0,0 +1,116 @@
+ *   attacks).
+ * - set response headers to prevent certain kinds of attacks.
+ */
+private class HttpSecurityFilter(
+    conf: SparkConf,
+    securityMgr: SecurityManager) extends Filter {
+
+  private val NEWLINE_AND_SINGLE_QUOTE_REGEX = raw"(?i)(\r\n|\n|\r|%0D%0A|%0A|%0D|'|%27)".r
+
+  override def destroy(): Unit = { }
+
+  override def init(config: FilterConfig): Unit = { }
+
+  override def doFilter(req: ServletRequest, res: ServletResponse, chain: FilterChain): Unit = {
+    val hreq = req.asInstanceOf[HttpServletRequest]
+    val hres = res.asInstanceOf[HttpServletResponse]
+    hres.setHeader("Cache-Control", "no-cache, no-store, must-revalidate")
+
+    if (!securityMgr.checkUIViewPermissions(hreq.getRemoteUser())) {
+      hres.sendError(HttpServletResponse.SC_FORBIDDEN,
+        "User is not authorized to access this page.")
+      return
+    }
+
+    // Check if any disallowed content is in the incoming headers or parameters. This filters
+    // out content that could be used for XSS attacks from even making it to the UI handlers.
+    try {
+      hreq.getHeaderNames().asScala.foreach { k =>
+        require(isSafe(k), "Request header name contains disallowed content.")
+        hreq.getHeaders(k).asScala.foreach { v =>
+          require(isSafe(v), s"Header value for $k contains disallowed content.")
+        }
+      }
+
+      hreq.getParameterMap().asScala.foreach { case (k, values) =>
+        require(isSafe(k), "Parameter name contains disallowed content.")
+        values.foreach { v =>
+          require(isSafe(v), s"Parameter value for $k contains disallowed content.")
+        }
+      }
+    } catch {
+      case e: IllegalArgumentException =>
+        // Do not allow the request to go further if any problem is detected.
+        hres.setContentType("text/plain;charset=utf-8")
+        hres.sendError(HttpServletResponse.SC_BAD_REQUEST, e.getMessage())
+        return
+    }
+
+    // SPARK-10589 avoid frame-related click-jacking vulnerability, using X-Frame-Options
+    // (see http://tools.ietf.org/html/rfc7034). By default allow framing only from the
+    // same origin, but allow framing for a specific named URI.
+    // Example: spark.ui.allowFramingFrom = https://example.com/
+    val xFrameOptionsValue = conf.getOption("spark.ui.allowFramingFrom")
+      .map { uri => s"ALLOW-FROM $uri" }
+      .getOrElse("SAMEORIGIN")
+
+    hres.setHeader("X-Frame-Options", xFrameOptionsValue)
+    hres.setHeader("X-XSS-Protection", conf.get(UI_X_XSS_PROTECTION))
+    if (conf.get(UI_X_CONTENT_TYPE_OPTIONS)) {
+      hres.setHeader("X-Content-Type-Options", "nosniff")
+    }
+    if (hreq.getScheme() == "https") {
+      conf.get(UI_STRICT_TRANSPORT_SECURITY).foreach(
+        hres.setHeader("Strict-Transport-Security", _))
+    }
+
+    chain.doFilter(req, res)
+  }
+
+  private def isSafe(str: String): Boolean = stripXSS(str) == str
```

Comment:
we didn't used to have this check.  Are you sure this is safe everywhere?  Eg. before if a param was `"runtime>10ms"` it would become `"runtime&gt;10ms"`, and with this change, it'll fail

Reply:
I'll take a look, but I don't remember any pages that use that sort of syntax.

(I did find a separate issue with this filtering that I'll fix.)

Reply:
It seems that while no pages I found use that syntax, the search boxes in the fancy task page (and probably other places) don't like this if you type an unexpected character.

I'll have to think about how to fix that and still keep this logic centralized...

Reply:
I see, the reason you don't just update the params is that the `HttpServletRequest` doesn't let you?

The only workaround I can think of is to stuff the modified values into an attribute of the ServletContext, and then have a style rule to ban directly accessing them.

Reply:
I can use an `HttpServletRequestWrapper` to do that, I just thought it would be better (and easier) to just error out instead of modifying the request. I'll go that route instead, at least we know the behavior will be the same and won't break something inadvertently...